src/tutorials/boss_tutorial.md

```diff
@@ -0,0 +1,4330 @@
+            end
+        end
+    end
+    ```
+
+    The `OnImpact` will look at the object being impacted to see if it is invulnerable. If it is, the owner of the weapon receives a broadcast that will show the damage. The 4th argument is set to true, indicating the object hit is immune to the damage.
+
+### Test the game
+
+When a player shoots the boss, it will display **Immune**.
+
+### Updated PlayerServer Script
+
+??? "PlayerServer"
+    ```lua
+    local WEAPON = script:GetCustomProperty("Weapon")
+
+    -- Store a list of players in the game so that
+    -- the events can be disconnected later when
+    -- they leave.
+    local players = {}
+
+    -- Enable weapon abilities for the player
+    local function EnableWeapon(player)
+        if Object.IsValid(player) and players[player.id] and Object.IsValid(players[player.id].weapon) then
+            local abilities = players[player.id].weapon:FindDescendantsByType("Ability")
+
+            for index, ability in ipairs(abilities) do
+                ability.isEnabled = true
+            end
+        end
+    end
+
+    -- Disable weapon abilities for the player
+    local function DisableWeapon(player)
+        if Object.IsValid(player) and players[player.id] and Object.IsValid(players[player.id].weapon) then
+            local abilities = players[player.id].weapon:FindDescendantsByType("Ability")
+
+            for index, ability in ipairs(abilities) do
+                ability.isEnabled = false
+            end
+        end
+    end
+
+    -- Look for a valid target that is a damageable object
+    local function GetValidTarget(target)
+        if not Object.IsValid(target) then
+            return nil
+        end
+
+        if target:IsA("Damageable") then
+            return target
+        end
+
+        return target:FindAncestorByType("Damageable")
+    end
+
+    -- See if the target is invulnerable, if so, broadcast
+    -- to the player for feedback.
+    local function OnImpact(weaponObj, impactData)
+        local target = GetValidTarget(impactData.targetObject)
+
+        if Object.IsValid(target) then
+            if target.isInvulnerable then
+                Events.BroadcastToPlayer(impactData.weaponOwner, "ShowDamage", 0, true)
+            end
+        end
+    end
+
+    -- When a player joins, give a weapon
+    local function OnPlayerJoined(player)
+        local weapon = World.SpawnAsset(WEAPON)
+
+        weapon:Equip(player)
+        weapon.targetImpactedEvent:Connect(OnImpact)
+
+        players[player.id] = {
+
+            weapon = weapon
+
+        }
+
+        DisableWeapon(player)
+    end
+
+    local function OnPlayerLeft(player)
+        players[player.id] = nil
+    end
+
+    Game.playerJoinedEvent:Connect(OnPlayerJoined)
+    Game.playerLeftEvent:Connect(OnPlayerLeft)
+
+    Events.Connect("EnableWeapon", EnableWeapon)
+    Events.Connect("DisableWeapon", DisableWeapon)
+    ```
+
+## Add Player Sprint Ability
```

Comment:
At the h2, go back to known or gerund. "Adding Player Sprint ... ", "Giving Your Players The Ability to Sprint".

It needs to sound more general than the h3 to help it break up the sections semantically.